
```{r setup, include=FALSE, echo=FALSE, include=FALSE}
### Set knitr global options
knitr::opts_chunk$set(warning=FALSE, message=FALSE)

### Fix conflicts
conflict_prefer("filter", "dplyr")
conflict_prefer("count", "dplyr")
conflict_prefer("select", "dplyr")
conflict_prefer("slice", "dplyr")
conflict_prefer("rename","dplyr")
conflict_prefer("mutate","dplyr")
conflict_prefer("arrange","dplyr")
conflict_prefer("summarize", "dplyr")
conflict_prefer("summarise", "dplyr")
conflict_prefer("layout","plotly")
```

```{r, echo = FALSE, include = FALSE}
if(!exists("seurats")){
  seurats = readRDS(params$qcs_save_name)
}
```

## 2.1 - Annotate with singleR

```{r, include=FALSE}
anno_singleR_text = paste0("Reference data are load with Ensembl annotations.
`ensembl = F` indicates to not convert row names to ensembl IDs.
The selected annotation package is ", params$anno_annotation)

if(!as.logical(params$anno_singleR)){
  anno_singleR_text = "> **_NOTE:_** The file has been already labeled by QuantBio with SingleR, and QC checks have been performed on the annotations. This step has been skipped."
}
```

This subsection is **code only**. Hit the "code" button to reveal the code. `r anno_singleR_text`

```{r anno_singleR, tidy='styler', tidy.opts=list(strict=FALSE), eval = params$anno_singleR}
if(params$anno_annotation == "HumanAtlas"){
  ref.data <- HumanPrimaryCellAtlasData(ensembl=FALSE)   
}else if(params$anno_annotation == "ImmGenData"){
  ref.data <- ImmGenData(ensembl=FALSE)  
}


### Assign the cell types. For this step, I'm parellelizing it with 
### the BiocParallel library. Parallelizing may not be necessary or helpful,
### but its here. 
cells <- lapply(X=seurats, FUN=function(x) {
    SingleR(test=LayerData(x, assay = "RNA", layer = "counts"),
                 ref=ref.data, labels=ref.data$label.main,
                 BPPARAM=MulticoreParam(detectCores())) 
})

### Add to seurat
for(i in 1:length(seurats)) {
    seurats[[i]]$SingleR <- cells[[i]]$labels
}
```

## 2.2 - Perform QC checks on new cell labels {.tabset .tabset-pils}

### 1 - Heatmap before pruning {.tabset .tabset-pils}

Here we prune out cell labels with low quality cell-type-assignment.
Low delta values are caused by:

1. ambiguous assignments with closely related reference labels
2. incorrect assignments that match poorly to all reference labels.

We first check the scores and the number of low-quality labels

```{r anno_qc1, tidy='styler', tidy.opts=list(strict=FALSE), warning=FALSE,  fig.align='center',fig.height=8, fig.width = 14, fig.cap="All cell types identified and their scores (pre pruned)", eval = params$anno_singleR}
### Generate a list of plots based on the above information
plots<-list()
for (i in 1:length(seurats)) {
plotScoreHeatmap(cells[[i]], main = names(seurats[i]))
}
```
<br>
<br>

### 2 - Delta score before pruning {.tabset .tabset-pils}

```{r anno_qc2, results="asis",fig.height=10, fig.width = 10, warning=FALSE, fig.cap = "Number of low-quality labels per cell assignment (pre pruned)", fig.align='center', tidy='styler', tidy.opts=list(strict=FALSE), eval = params$anno_singleR}
### Generate a list of plots based on the above information
plots<-list()
for (i in 1:length(seurats)) {
    p1 <- eval(substitute(
        
        plotDeltaDistribution(cells[[i]],
                              ncol = 5,
                              size=0.5) +
            themes$QB_theme()+
            ylab("Delta score") +
            xlab("Cell labels") +
            ggtitle(names(seurats[i]))

        ,list(i = i)))
    plots[[i]] <- p1  # add each plot into plot list
}

#### Plot it out
for (i in 1:length(plots)){
print(glue::glue("#### {as.vector(names(seurats[i]))}\n"))
  
title_plots <- plots[[i]]
print(title_plots)
cat("\n \n \n")
}
```


<br>
<br>

### 3 - Count of cells to be pruned {.tabset .tabset-pils}

**Table of pruned cell counts:**

```{r results = "asis", anno_qc3, tidy='styler', tidy.opts=list(strict=FALSE), eval=params$anno_singleR}
tabs<-list()
for(i in 1:length(cells)) {
    tabs[[i]]<-table(table(is.na(cells[[i]]$pruned.labels)) |>
        as.data.frame() |>
        dplyr::mutate(Group = names(seurats[i])) |>
        dplyr::rename(Pruned = Var1, Count = Freq)) 
    
        tabs[[i]] <- as.data.frame(tabs[[i]]) |>
        dplyr::filter(Freq>0)|>
        dplyr::select(Group, Pruned, Count)
        tabs[[i]] <- tabs[[i]] |> render_dt()
            
}


#### Plot it
for (i in 1:length(tabs)){
print(glue::glue("#### {as.vector(names(seurats[i]))}\n"))
  
title_plots <- tabs[[i]]
print(title_plots)
cat("\n \n \n")
}

```

<br>
<br>


```{r tidy='styler', tidy.opts=list(strict=FALSE),eval=params$anno_singleR}
## Prune out the low-quality reads
prune<-list()
keep<-list()
for(i in 1:length(cells)){
    prune[[i]]<-pruneScores(cells[[i]],
                   nmads = 3,
                   min.diff.med = -Inf,
                   min.diff.next = 0,
                   get.thresholds = FALSE
                   )
    prune[[i]]<-which(prune[[i]]=="TRUE")
    keep[[i]]<-cells[[i]][-prune[[i]],]
}
```

### 4 - Heatmap after pruning

```{r anno_qc1, fig.height=8, fig.width = 10, fig.cap="All cell types identified and their scores (after prune)", fig.align='center', warning=FALSE, message=FALSE, tidy='styler', tidy.opts=list(strict=FALSE),eval=params$anno_singleR}
### Re-plot to ensure prune worked
plots<-list()
for (i in 1:length(seurats)) {
plotScoreHeatmap(keep[[i]], main = names(seurats[i]))
}
```

<br>
<br>

### 5 - Delta score after pruning {.tabset .tabset-pils}

```{r anno_qc2, results="asis", fig.height=8, fig.width = 10, fig.align='center', fig.cap = "Number of low-quality labels per cell assignment (after prune)", warning=FALSE,  tidy='styler', tidy.opts=list(strict=FALSE),eval=params$anno_singleR}
plots<-list()
for (i in 1:length(seurats)) {
    p1 <- eval(substitute(
        
        plotDeltaDistribution(keep[[i]],
                              ncol = 5,
                              size=0.5) +
            themes$QB_theme()+
            ylab("Delta score") +
            xlab("Cell labels") +
            ggtitle(names(seurats[i]))

        ,list(i = i)))
    plots[[i]] <- p1  # add each plot into plot list
}

#### Plot it out
for (i in 1:length(plots)){
print(glue::glue("#### {as.vector(names(seurats[i]))}\n"))
  
title_plots <- plots[[i]]
print(title_plots)
cat("\n \n \n")
}
```

<br>
<br>

### 6 - Prune cells from seurat

This subsection is **code only**. Hit the "code" button to reveal the code.

```{r anno_pruned, tidy='styler', tidy.opts=list(strict=FALSE),eval=TRUE}
### Prune the cells in the subsetted seurat objects
for(i in 1:length(seurats)){
    seurats[[i]]<-seurats[[i]][,-prune[[i]]]
    seurats[[i]]$SingleR <- factor(seurats[[i]]$SingleR)
    }
```

<br>
<br>

## 2.4 - Merge similar cell types

------------------------------------------------------------------------

Merge ambiguous cell types into:

(1) Fibroblasts 
(2) Macrophages
(3) B_cells
(4) Others

If the cells are below **`r params$anno_cutoff`**, they are merge to "other". 

**NOTE:**

Neurons are also added to this group because neurons aren't an anticipated cell type. 
Also, if the cells are B or T, cell subtypes merge with B and T cells, respectively.


```{r merge_cell_types, tidy='styler', tidy.opts=list(strict=FALSE),eval=params$anno_singleR}
### So not to lose the low count b cell subtypes, collapse those first
freqs<-list()
others<-list()
metas<-list()
for(i in 1:length(seurats)){
    seurats[[i]]$SingleR2 <- fct_collapse(seurats[[i]]$SingleR,
                                         B_cell = params$anno_Bcells,
                                         HSC = params$anno_HCS,
                                         Fibroblasts = params$anno_fibroblast)
    
    ### Generate frequency tables
    metas[[i]]<-seurats[[i]]@meta.data
    freqs[[i]]<-data.frame(tabyl(metas[[i]]$SingleR2, sort = TRUE))
    colnames(freqs[[i]])<-c("Cells","n","Proportion")
    freqs[[i]]$Percent<-freqs[[i]]$Proportion*100
    freqs[[i]]<-freqs[[i]][order(freqs[[i]]$n, decreasing = TRUE),]
    
    ### I used the above "tabyl" function to get the proportion of cells out of 
    ### all total cells in case we want to set a cut-off using a proportion rather
    ### than total count. 
    others[[i]]<-as.vector(freqs[[i]]$cells[which(freqs[[i]]$n<params$anno_ctoff)])
    seurats[[i]]$SingleR2 <- fct_collapse(seurats[[i]]$SingleR2,
                                                       Others=c(others,
                                                                params$anno_Others))
    Idents(seurats[[i]])<-seurats[[i]]$SingleR2
    }
```

### Count cells types {.tabset .tabs-pils}

```{r, echo = FALSE, results="asis", tidy='styler', tidy.opts=list(strict=FALSE),eval=params$anno_singleR}
### Generate frequency tables
tabs<-list()
for (i in 1:length(seurats)) {
#    p1 <- eval(substitute(
        
        tabs[[i]]<-tabyl(seurats[[i]]@meta.data$SingleR2) |>  as.data.frame() 

        colnames(tabs[[i]])<-c("Cell type","n","Proportion of cells") 
        tabs[[i]]$Group<-names(seurats[i])
        
        
            # dplyr::rename(`Cell type` = {{paste0("seurats[[","L",i,"]]@meta.data$SingleR2")}},
            #               `Number of cells` = n, 
            #               Percent = percent) #|>
            # tabs[[i]] |> knitr::kable( caption = "Proportion of cells",
            #              digits = 2
            #              )

#        ,list(i = i)))
#    tabs[[i]] <- p1  # add each plot into plot list
}

#### Print them
stacked_dataframe <- do.call(rbind, tabs) |> 
    select(Group, 
           `Cell type`,
           n,
           `Proportion of cells`) 

### Table it
stacked_dataframe |> render_dt()
```

<br>
<br>

## 2.5 - Identify tumor cells

This subsection is **code only**. Hit the "code" button to reveal the code.This step utilizes `inferCNV` to infer copy number variations that will allow us to identify malignant cells.

> **_NOTE:_** For Now Skip

```{r, eval=FALSE, include=FALSE}
### For inferCNV, we will have to assemble the inferCNV object which requires the
### use of the following 3 objects: 1 - expression matrix, 2 - annotation matrix, 
### and 3 - go (gene order) matrix. 

### 1 - EXPRESSION MATRIX: Due to the length it will take to run the full
### Seurat object, we will subset the seurat object to have no more than a
### maximum of 2,000 cells per cell type, then run inferCNV. Downsampling
### can be performed here or when assembling the inferCNV object. Due to the
### way we will build the annotation matrix and go matrix and the inferCNV
### object, I'll do it here. If done downstream, it gets complicated.
temp<-seurat_analysed
rc<-as.matrix(temp@assays$RNA@data)

### 2 - ANNOTATION MATRIX
anns<-data.frame(labels=temp@meta.data[,c("SingleR2")])
rownames(anns)<-rownames(temp@meta.data)

### 3 - GO MATRIX: This matrix has the chromosome name, start, and end position
### Since it is not present in Seurat, I will have to build it with biomart
go<-data.frame(genes=rownames(temp),
                  chromosome="",
                  start="",
                  end="")
ensembl <- useEnsembl(biomart = "genes")
ensembl <- useDataset(dataset = "hsapiens_gene_ensembl", mart = ensembl)

### look at filters
filters = listFilters(ensembl)
attributes = listAttributes(ensembl)

### Get the chromosome information
chromosome_stuff<-getBM(attributes = c('entrezgene_accession',
                                       'chromosome_name',
                                       'start_position',
                                       'end_position'),
      filters = 'entrezgene_accession', 
      values = go$genes, 
      mart = ensembl)

go$chromosome <- mapvalues(x=go$gene, 
                         from=chromosome_stuff$entrezgene_accession, 
                         to=as.character(as.vector(chromosome_stuff$chromosome_name)))

go$start <- mapvalues(x=go$gene, 
                         from=chromosome_stuff$entrezgene_accession, 
                         to=as.character(as.vector(chromosome_stuff$start_position)))

go$start<-as.numeric(go$start)

go$end <- mapvalues(x=go$gene, 
                         from=chromosome_stuff$entrezgene_accession, 
                         to=as.character(as.vector(chromosome_stuff$end_position)))

go$end<-as.numeric(go$end)

### Subset by bonafide chromosome assignments (other chromosomes are haplotypes)
### I'm sure there's a faster way to do this, just haven't found out
go<-go[which(go$chromosome==1|go$chromosome==2|
                 go$chromosome==3|go$chromosome==4|
                 go$chromosome==5|go$chromosome==6|
                 go$chromosome==7|go$chromosome==8|
                 go$chromosome==9|go$chromosome==10|
                 go$chromosome==11|go$chromosome==12|
                 go$chromosome==13|go$chromosome==14|
                 go$chromosome==15|go$chromosome==16|
                 go$chromosome==17|go$chromosome==18|
                 go$chromosome==19|go$chromosome==20|
                 go$chromosome==21|go$chromosome==22),]

rownames(go)<-go$genes
go$genes<-NULL

### 4 - inferCNV OBJECT: Now assemble the inferCNV object
infercnv_obj = CreateInfercnvObject(raw_counts_matrix = rc,
                                    annotations_file=anns,
                                    gene_order_file=go,     
                                    ref_group_names=c("T_cells","B_cell",
                                                      "Macrophage","Monocyte",
#                                                      "DC", "CMP", 
                                                      "Fibroblasts")
                                    ) 
  
### At long last, run inferCNV. For p-cut-offs, use 1 for smart-seq and 
### 0.1 for 10x-genomics
dir = "~/Desktop/inferCNV_output"
infercnv_obj = infercnv::run(infercnv_obj,
                             cutoff=0.1,  
                             out_dir=dir, 
                             cluster_by_groups=TRUE,  
                             denoise=TRUE,
                             num_threads=16, ### this PC has 16, check yours.
                             HMM=TRUE,
                             output_format = "pdf"#,
#                             png_res=150
                             )

### Add the results to the temporary seurat object
temp=infercnv::add_to_seurat(infercnv_output_path = dir,
                             seurat_obj = temp,
                             top_n=10)
temp_meta<-temp@meta.data

### Sum the total count of inferred CNVs
lets_see<-temp_meta[,grepl("has_cnv" ,colnames(temp_meta))]
lets_see$SingleR2<-temp_meta$SingleR2
lets_melt<-melt(lets_see,
                id.vars="SingleR2",
                variable.name="CNV",
                value.name="has_cnv")
cnv_counts<-table(lets_melt$SingleR2, lets_melt$has_cnv)
cnv_counts<-data.frame(no_cnv=cnv_counts[,1], has_cnv=cnv_counts[,2])

### Identify likely  malignant cells
malignant_cells<-rownames(cnv_counts[which.max(cnv_counts$has_cnv),])

```

<br>
<br>

### Replace the identified malignant cells types in the original seurat

```{r, include=FALSE, eval=FALSE}
seurat_analysed@meta.data$SingleR2 <- fct_collapse(seurat_analysed@meta.data$SingleR2,
                                           Tumor_cells=malignant_cells)
```

## 2.4 - Write to file
The annotation is done, the file is save on **_`r params$anno_save_name`_**.

```{r, include=FALSE, eval=params$anno_save,eval=FALSE}
### Merge it and write final object
write_rds(seurats, file=params$anno_save_name)
```
