
```{r setup, include=FALSE, echo=FALSE, include=FALSE}
### Set knitr global options
knitr::opts_chunk$set(warning=FALSE, message=FALSE)

### Load libraries
library(data.table)
library(SingleR)
library(Seurat)
library(glmnet)
library(forcats)
# library(infercnv)
library(biomaRt)
library(plyr)
library(ggplot2)
library(janitor)
library(BiocParallel)
library(readr)
library(celldex)
library(viridis)
library(pheatmap)
require(styler)

theme_sara <- function() {
  theme_bw(base_size=14)+
    theme(axis.text=element_text(color="black"),
          panel.background=element_rect(color="black"),
          strip.text = element_text(size=12),
          strip.background = element_rect(fill="white"))
}
```

```{r, echo = FALSE, include = FALSE}
if(!exists("seurat_analysed")){
  seurat_analysed = readRDS(params$qcs_save_file)
}
```

## 2.1 - Annotate with singleR

```{r, include=FALSE}
anno_sigleR_text = "Reference data are load with Ensembl annotations.
`ensembl = F` indicates to not convert row names to ensembl IDs."

if(!as.logical(params$anno_sigleR)){
  anno_sigleR_text = "> **_NOTE:_** The file has been already labeled by QuantBio with SingleR, and QC checks have been performed on the annotations. This step has been skipped."
}
```
`r anno_sigleR_text`

```{r anno_singleR, tidy='styler', tidy.opts=list(strict=FALSE), eval = params$anno_sigleR}
ref.data <- HumanPrimaryCellAtlasData(ensembl=FALSE) 

### Assign the cell types. For this step, I'm parellelizing it with 
### the BiocParallel library. Parallelizing may not be necessary or helpful,
### but its here. 
cells <- SingleR(test=seurat_analysed@assays[["RNA"]]@data,
                 ref=ref.data, labels=ref.data$label.main,
                 BPPARAM=MulticoreParam(detectCores())) 

### Add to seurat
seurat_analysed@meta.data$SingleR <- cells$labels
```

## 2.2 - Perform QC checks on new cell labels

Here we prune out cell labels with low quality cell-type-assignment.
Low delta values are caused by:

1. ambiguous assignments with closely related reference labels
2. incorrect assignments that match poorly to all reference labels.

We first check the scores and the number of low-quality labels

```{r anno_qc1, fig.height=8, fig.width = 10, fig.cap="All cell types identified and their scores (pre pruned)", fig.align='center', tidy='styler', tidy.opts=list(strict=FALSE), warning=FALSE}
plotScoreHeatmap(cells)
```

```{r anno_qc2, fig.height=8, fig.width = 10, warning=FALSE, fig.cap = "Number of low-quality labels per cell assignment (pre pruned)", fig.align='center', tidy='styler', tidy.opts=list(strict=FALSE)}
plotDeltaDistribution(cells, 
                      ncol = 5,
                      size=0.5) +
  theme_sara()+
  ylab("Delta score") +
  xlab("Cell labels")

```

```{r anno_qc3, tidy='styler', tidy.opts=list(strict=FALSE)}
table(is.na(cells$pruned.labels)) %>%
  as.data.frame() %>%
  dplyr::rename(Pruned = Var1, Count = Freq) %>%
  knitr::kable(.,
             caption = "Count the cells that need to be pruned")
```

We prune **`r as.numeric(table(is.na(cells$pruned.labels))["TRUE"])` cells**.
```{r tidy='styler', tidy.opts=list(strict=FALSE)}
## Prune out the low-quality reads
prune<-pruneScores(cells,
                   nmads = 3,
                   min.diff.med = -Inf,
                   min.diff.next = 0,
                   get.thresholds = FALSE
                   )

prune<-which(prune=="TRUE")
keep<-cells[-prune,]
```

```{r anno_qc1, fig.height=8, fig.width = 10, fig.cap="All cell types identified and their scores (after prune)", fig.align='center', warning=FALSE, message=FALSE, tidy='styler', tidy.opts=list(strict=FALSE)}

### Re-plot to ensure prune worked
plotScoreHeatmap(keep) 
```

```{r anno_qc2, fig.height=8, fig.width = 10, fig.align='center', fig.cap = "Number of low-quality labels per cell assignment (after prune)", warning=FALSE,  tidy='styler', tidy.opts=list(strict=FALSE)}

plotDeltaDistribution(keep, 
                      ncol = 5,
                      size=0.5) +
  theme_sara() +
  ylab("Delta score") +
  xlab("Cell labels")
```

## 2.3 - Prune ambiguous cells

```{r anno_pruned, tidy='styler', tidy.opts=list(strict=FALSE)}
### Prune the cells in the subsetted seurat objects
seurat_analysed<-seurat_analysed[,-prune]

### Convert to factor
seurat_analysed@meta.data$SingleR <- factor(seurat_analysed@meta.data$SingleR)
# levels(seurat_analysed@meta.data$SingleR)
```

## 2.4 - Merge similar cell types

Merge ambiguous cell types into:

(1) Fibroblasts 
(2) Macrophages
(3) B_cells
(4) Others

If the cells are below **`r params$anno_cutoff`**, they are merge to "other". 

***
**NOTE:**

Neurons are also added to this group because neurons aren't an anticipated cell type. 
Also, if the cells are B or T, cell subtypes merge with B and T cells, respectively.

***

```{r merge_cell_types, tidy='styler', tidy.opts=list(strict=FALSE)}
### So not to lose the low count b cell subtypes, collapse those first
seurat_analysed@meta.data$SingleR2 <- fct_collapse(seurat_analysed@meta.data$SingleR,
                                           B_cell = params$anno_BCell,
                                           HSC = params$anno_HCS,
                                           Fibroblasts = params$anno_fibroblast)

freq<-data.frame(tabyl(seurat_analysed@meta.data$SingleR2, sort = TRUE))
colnames(freq)<-c("cells","n","proportion")
freq$percent<-freq$proportion*100
freq<-freq[order(freq$n, decreasing = TRUE),]

### I used the above "tabyl" function to get the proportion of cells out of 
### all total cells in case we want to set a cut-off using a proportion rather
### than total count. 
others<-as.vector(freq$cells[which(freq$n<params$anno_cutoff)])

seurat_analysed@meta.data$SingleR2 <- fct_collapse(seurat_analysed@meta.data$SingleR2,
                                           Others=c(others,
                                                    params$anno_Others))

### Change the identity from tissue type to cell type 
Idents(seurat_analysed)<-seurat_analysed@meta.data$SingleR2
```

```{r, echo = FALSE, tidy='styler', tidy.opts=list(strict=FALSE)}
tabyl(seurat_analysed@meta.data$SingleR2) %>%
  as.data.frame() %>%
  dplyr::rename(`Cell type` = `seurat_analysed@meta.data$SingleR2`, 
                `Number of cells` = n, Percent = percent) %>%
    knitr::kable(.,
             caption = "Proportion of cells",
             digits = 2)
```

## 2.5 - Identify tumor cells

This step utilizes `inferCNV` to infer copy number variations that will allow
us to identify malignant cells.

> **_NOTE:_** For Now Skip

```{r, eval=FALSE, include=FALSE}
### For inferCNV, we will have to assemble the inferCNV object which requires the
### use of the following 3 objects: 1 - expression matrix, 2 - annotation matrix, 
### and 3 - go (gene order) matrix. 

### 1 - EXPRESSION MATRIX: Due to the length it will take to run the full
### Seurat object, we will subset the seurat object to have no more than a
### maximum of 2,000 cells per cell type, then run inferCNV. Downsampling
### can be performed here or when assembling the inferCNV object. Due to the
### way we will build the annotation matrix and go matrix and the inferCNV
### object, I'll do it here. If done downstream, it gets complicated.
temp<-seurat_analysed
rc<-as.matrix(temp@assays$RNA@data)

### 2 - ANNOTATION MATRIX
anns<-data.frame(labels=temp@meta.data[,c("SingleR2")])
rownames(anns)<-rownames(temp@meta.data)

### 3 - GO MATRIX: This matrix has the chromosome name, start, and end position
### Since it is not present in Seurat, I will have to build it with biomart
go<-data.frame(genes=rownames(temp),
                  chromosome="",
                  start="",
                  end="")
ensembl <- useEnsembl(biomart = "genes")
ensembl <- useDataset(dataset = "hsapiens_gene_ensembl", mart = ensembl)

### look at filters
filters = listFilters(ensembl)
attributes = listAttributes(ensembl)

### Get the chromosome information
chromosome_stuff<-getBM(attributes = c('entrezgene_accession',
                                       'chromosome_name',
                                       'start_position',
                                       'end_position'),
      filters = 'entrezgene_accession', 
      values = go$genes, 
      mart = ensembl)

go$chromosome <- mapvalues(x=go$gene, 
                         from=chromosome_stuff$entrezgene_accession, 
                         to=as.character(as.vector(chromosome_stuff$chromosome_name)))

go$start <- mapvalues(x=go$gene, 
                         from=chromosome_stuff$entrezgene_accession, 
                         to=as.character(as.vector(chromosome_stuff$start_position)))

go$start<-as.numeric(go$start)

go$end <- mapvalues(x=go$gene, 
                         from=chromosome_stuff$entrezgene_accession, 
                         to=as.character(as.vector(chromosome_stuff$end_position)))

go$end<-as.numeric(go$end)

### Subset by bonafide chromosome assignments (other chromosomes are haplotypes)
### I'm sure there's a faster way to do this, just haven't found out
go<-go[which(go$chromosome==1|go$chromosome==2|
                 go$chromosome==3|go$chromosome==4|
                 go$chromosome==5|go$chromosome==6|
                 go$chromosome==7|go$chromosome==8|
                 go$chromosome==9|go$chromosome==10|
                 go$chromosome==11|go$chromosome==12|
                 go$chromosome==13|go$chromosome==14|
                 go$chromosome==15|go$chromosome==16|
                 go$chromosome==17|go$chromosome==18|
                 go$chromosome==19|go$chromosome==20|
                 go$chromosome==21|go$chromosome==22),]

rownames(go)<-go$genes
go$genes<-NULL

### 4 - inferCNV OBJECT: Now assemble the inferCNV object
infercnv_obj = CreateInfercnvObject(raw_counts_matrix = rc,
                                    annotations_file=anns,
                                    gene_order_file=go,     
                                    ref_group_names=c("T_cells","B_cell",
                                                      "Macrophage","Monocyte",
#                                                      "DC", "CMP", 
                                                      "Fibroblasts")
                                    ) 
  
### At long last, run inferCNV. For p-cut-offs, use 1 for smart-seq and 
### 0.1 for 10x-genomics
dir = "~/Desktop/inferCNV_output"
infercnv_obj = infercnv::run(infercnv_obj,
                             cutoff=0.1,  
                             out_dir=dir, 
                             cluster_by_groups=TRUE,  
                             denoise=TRUE,
                             num_threads=16, ### this PC has 16, check yours.
                             HMM=TRUE,
                             output_format = "pdf"#,
#                             png_res=150
                             )

### Add the results to the temporary seurat object
temp=infercnv::add_to_seurat(infercnv_output_path = dir,
                             seurat_obj = temp,
                             top_n=10)
temp_meta<-temp@meta.data

### Sum the total count of inferred CNVs
lets_see<-temp_meta[,grepl("has_cnv" ,colnames(temp_meta))]
lets_see$SingleR2<-temp_meta$SingleR2
lets_melt<-melt(lets_see,
                id.vars="SingleR2",
                variable.name="CNV",
                value.name="has_cnv")
cnv_counts<-table(lets_melt$SingleR2, lets_melt$has_cnv)
cnv_counts<-data.frame(no_cnv=cnv_counts[,1], has_cnv=cnv_counts[,2])

### Identify likely  malignant cells
malignant_cells<-rownames(cnv_counts[which.max(cnv_counts$has_cnv),])

```

### Replace the identified malignant cells types in the original seurat
> **_NOTE:_** For Now Skip

```{r, include=FALSE, eval=FALSE}
seurat_analysed@meta.data$SingleR2 <- fct_collapse(seurat_analysed@meta.data$SingleR2,
                                           Tumor_cells=malignant_cells)
```

## 2.4 - Write to file
The annotation is done, the file is save on **_`r params$anno_save_file`_**.

```{r, include=FALSE}
write_rds(seurat_analysed, file=params$anno_save_file)
```
