---
title: "Bassez breast scRNA-seq analysis"
author: "QuantBio - VMP"
date: 9-30-2021
output:
  html_document:
    toc: yes
    toc_float: yes
    number_sections: yes
  pdf_document:
    toc: yes
    toc_depth: 4
---

ROSA - Things to parameterize, if possible:
  [1] - Line 152, where I set the stdev cut-off for the PCs selected
  [2] - Line 244 and 245, where we set the cut-offs for pct and fold-change

  ALSO - this file may take longer to run than the other files due to
  differential expression. I recommend downsampling before you start using:
  Idents(bassez1)<-bassez1@meta.data$SingleR2
  bassez1<-subset(bassez1, downsample=1000)

# Purpose
To identify differentiallly expressed genes in scRNA-seq data

```{r setup, include=FALSE, echo=FALSE}
### Set knitr global options
knitr::opts_chunk$set(echo = TRUE, warning=FALSE, message=FALSE)

### Load libraries
library(data.table)
library(ComplexHeatmap)
library(annotate)
library(SingleR)
library(celldex)
library(Seurat)
library(scales)
library(RColorBrewer)
library(ggplot2)
library(plyr)
library(kableExtra)
library(knitr)
library(forcats)
library(readxl)
library(dplyr)
library(ggpubr)
library(circlize)
library(readr)

### Establish color schemes and themes needed
theme_sara <- function() {
  theme_bw(base_size=14)+
    theme(axis.text=element_text(color="black"),
          panel.background=element_rect(color="black"),
          strip.text = element_text(size=12),
          strip.background = element_rect(fill="white"))
}

theme_sara_90 <- function() {
  theme_bw(base_size=18)+
    theme(axis.text.x=element_text(angle=90,hjust = 1,vjust = 0.5),
          axis.text=element_text(color="black"),
          panel.background=element_rect(color="black"),
          strip.text = element_text(size=12),
          strip.background = element_rect(fill="white"))
}

ct.colors <- c("T_cells"="chartreuse4",
               "B_cell"="darkorange",
               "NK_cell"="red",
               "Epithelial_cells"="firebrick", 
               "Endothelial_cells"="hotpink", 
               "Fibroblasts"="royalblue", 
               "Macrophage"="chocolate4", 
               "CMP"="navy",
               "Others"="darkorchid",
               "DC"="cyan4",
               "Monocyte"="bisque3")

cols <- colorRamp2( c(0.0,0.5,8.25), c("white","orangered2","black"))

phase.colors <- c("G1"="#BF6ECA",
               "G2M"="#ACDE7C",
               "S"="#BEBEBB")

### If more colors are needed, use the following-------------------------------
#library(randomcoloR)
#n <- 3
#palette <- distinctColorPalette(n)
#pie(rep(1, n), col=palette)
```

# Load the data

```{r}
bassez1<-read_rds(file="~/Dropbox (QuantBio LLC)/QB_shared/RESOURCES/scRNA-seq/Breast_Bassez_2021/Seurat_Breast_Bassez1_MINIATURE2.rds")

### Chang identity to cell type
Idents(bassez1)<-bassez1@meta.data$SingleR2
```

# Run PCA analyses

```{r}
### Perform linear dimensional reduction via PCA
bassez1 <- RunPCA(bassez1)

###----------------------------------------------------------------------------
### Run diagnostics on PCs
### Method 1 - Print the PCs
print(bassez1[["pca"]], dims = 1:15, nfeatures = 5)

### Method 2 - VizDimLoadings
### This plot is to identify genes contributing to desired PCAs
VizDimLoadings(bassez1, 
               dims = 1:4, 
               reduction = "pca")

### Method 3 - Dimplot
DimPlot(bassez1, reduction = "pca",
        cols = ct.colors,
        pt.size = 0.5) +
  ggtitle("PCA of cell types defined by SingleR") +
  theme_sara()+
  ylab("PC 2")+
  xlab("PC 1")+
  theme(axis.title.x = element_text())

FeaturePlot(bassez1,  
            reduction='pca', 
            features=c("nFeature_RNA",  "percent.mt"))

Idents(bassez1)<-bassez1@meta.data$Phase
DimPlot(bassez1, reduction = "pca",
        cols = phase.colors,
        pt.size = 0.5) +
  ggtitle("PCA of cells colored by cell cycle phase") +
  theme_sara()+
  ylab("PC 2")+
  xlab("PC 1")+
  theme(axis.title.y = element_text(),
        axis.title.x = element_text())
Idents(bassez1)<-bassez1@meta.data$SingleR2

### Method 4 - DimHeatmapt allows for easy exploration of the primary sources 
### of heterogeneity in a dataset, and can be useful when trying to decide which 
### PCs to include for further downstream analyses
DimHeatmap(bassez1, dims = 1:15, cells = 500, balanced = TRUE)

### Method 5 - Elbow plot
### We can observe an ‘elbow’ around PC6-7, , but even PCs up to 12 still maintain a high
### standard deviation. This is supported by the heatmaps. Suggesting that the majority of 
### true signal is captured in the first 12 PCs
ElbowPlot(bassez1)

### Identify most influential PCs
PCs<-which(bassez1@reductions$pca@stdev>=3)
```

# Plot uMAP and tSNE
NOTES: It is encouraged users to repeat downstream analyses with a different number of 
PCs (10, 15, or even 50!). As the results often do not differ dramatically. It is also 
advised users to err on the higher side when choosing this parameter. Because 
downstream analyses with only 5 PCs do significantly and adversely affect results.

Regarding clustering methods, SeuratV3 embeds cells in a graph structure - for example a 
K-nearest neighbor (KNN) graph, with edges drawn between cells with similar feature 
expression patterns, and then attempts to partition this graph into highly interconnected 
‘quasi-cliques’ or ‘communities’.

We first construct a KNN graph based on the euclidean distance in PCA space, and 
refine the edge weights between any two cells based on the shared overlap in their 
local neighborhoods (Jaccard similarity). This step is performed using the FindNeighbors()

To cluster the cells, we next apply modularity optimization techniques such as the 
Louvain algorithm (default) or SLM [SLM, Blondel et al., Journal of Statistical 
Mechanics], to iteratively group cells together, with the goal of optimizing the 
standard modularity function. The FindClusters() function implements this procedure.
Optimal resolution for 3K cells often ranges from 0.4 to 1.2 and increases the larger
the dataset.

```{r}
###--------------------THIS SECTION CAN BE SKIPPED----------------------------
###-----------UNLESS IT HASN'T BEEN DONE, THEN DON'T SKIP IT, DO IT-----------

### Identify neighbors
bassez1 <- FindNeighbors(bassez1, dims = 1:length(PCs))

### Identify clusters
bassez1 <- FindClusters(bassez1)

### Run uMAP and tSNE
bassez1 <- RunUMAP(bassez1, dims = 1:length(PCs))
bassez1 <- RunTSNE(bassez1, dims = 1:length(PCs))

### Plot uMAP 
DimPlot(bassez1, reduction = "umap",
        group.by = "SingleR2",
        cols = ct.colors,
#        shuffle = TRUE,
#        repel=TRUE,
#        label=TRUE, label.size = 2.5,label.box = TRUE,
        pt.size = 0.5) +
  ggtitle("UMAP of cell types defined by SingleR") +
  theme_sara()+
  ylab("UMAP 2")+
  xlab("UMAP 1")+
  theme(axis.title.y = element_text(),
        axis.title.x = element_text())

### Color by cell cycle phase
DimPlot(bassez1, reduction = "umap",
        group.by = "Phase",
        cols = phase.colors,
#        shuffle = TRUE,
#        repel=TRUE,
#        label=TRUE, label.size = 2.5,label.box = TRUE,
        pt.size = 0.5) +
  ggtitle("UMAP of cell types colored by cell cycle phase") +
  theme_sara()+
  ylab("UMAP 2")+
  xlab("UMAP 1")+
  theme(axis.title.y = element_text(),
        axis.title.x = element_text())
```


# Perform differential expression for each cluster
Seurat can help you find markers that define clusters via differential expression. 
By default, it identifies positive and negative markers of a single cluster (specified 
in ident.1), compared to all other cells. FindAllMarkers() automates this process for all 
clusters, but you can also test groups of clusters vs. each other, or against all cells.

The min.pct argument requires a feature to be detected at a minimum percentage in either 
of the two groups of cells, and the thresh.test argument requires a feature to be 
differentially expressed (on average) by some amount between the two groups. You can set 
both of these to 0, but with a dramatic increase in time - since this will test a large 
number of features that are unlikely to be highly discriminatory. As another option to 
speed up these computations, max.cells.per.ident can be set. This will downsample each 
identity class to have no more cells than whatever this is set to. While there is 
generally going to be a loss in power, the speed increases can be significant and the 
most highly differentially expressed features will likely still rise to the top.

```{r}
### Find markers for every cluster compared to all remaining cells
Idents(bassez1)<-bassez1@meta.data$SingleR2
DE_bassez1 <- FindAllMarkers(bassez1, 
#                             only.pos = TRUE, ### If you want only positive, use this.
                             min.pct = 0.25, 
                             logfc.threshold = 0.0)

```

# Identify significantly DE genes

```{r, fig.height=10, fig.width=10}

sigs <- DE_bassez1[which(DE_bassez1$p_val_adj<0.1 &
                           DE_bassez1$pct.1>0.25 &
                           DE_bassez1$avg_log2FC>0.0),]

### List the top 2 highest expressed genes for each cell subtype
sigs %>%
    group_by(cluster) %>%
    slice_max(n = 5, order_by = avg_log2FC)

sigs %>%
    group_by(cluster) %>%
    top_n(n = 5, wt = avg_log2FC) -> top5

```

# Plot heatmap of top 5 DEGs
Doheatmap() is no bueno, so using complexheatmap

```{r, fig.height=10, fig.width=12}
### Subset the top 5 genes
mat<-as.matrix(bassez1@assays[["RNA"]]@data)
mat<-mat[top5$gene,]

### Create an object to establish column annotations
ha <- HeatmapAnnotation(Celltype=bassez1@meta.data$SingleR2,
                        col=list(Celltype=ct.colors),
                        border=TRUE)

### Add row annotation colors
index<-match(top5$gene, DE_bassez1$gene)
ra <- rowAnnotation(`Fold-change (FC)`=anno_barplot(top5$avg_log2FC, 
                                               add_numbers=TRUE),
                    `FC cell reference` = top5$cluster,
                    annotation_name_rot=90,
                    col=list(`FC cell reference`=ct.colors),
                    gp = gpar(col = "black")
                    )

Heatmap(mat, 
        show_column_names = F, 
        col=cols, 
        top_annotation = ha, 
        column_split = bassez1@meta.data$SingleR2, 
        border = T, 
        cluster_column_slices = FALSE,
        cluster_row_slices =  FALSE,
        border_gp = gpar(col = "darkgrey", lty = 1),
        row_gap = unit(1.5, "mm"),
        row_split = top5$cluster,
        row_title = "DE Genes",
        column_title_side = "bottom", 
        column_title_rot = 90, 
        name="Expression", 
        left_annotation = ra
)
```

# Write files 

```{r}
### DEGs 
write.csv(sigs, file="~/Dropbox (QuantBio LLC)/QB_shared/RESOURCES/scRNA-seq/Breast_Bassez_2021/DEGs/DE_bassez1_MINATURE2.csv")

### Seurat
#write_rds(bassez1, file="~/Dropbox (QuantBio LLC)/QB_shared/RESOURCES/scRNA-seq/Breast_Bassez_2021/Seurat_Breast_Bassez1_MINIATURE2.rds")
```