---
title: "Single cell analysis: `r params$analysed_dataset`"
date: "Analysed on `r format(Sys.time(), '%B %d, %Y')`"
author: "QuantBio `r params$more_authors`"
output:
  rmdformats::downcute:
    self_contained: false
    thumbnails: false
    lightbox: true
    gallery: false
    highlight: tango
    code_folding: hide
    fig_width: 10
    fig_height: 10
    toc_depth: 5
params:
  analysed_dataset: "Example"
  more_authors: ""
  dir_seurat: "../../Dropbox (QuantBio LLC)/QB_shared/RESOURCES/scRNA-seq/Breast_Bassez_2021/Downloaded_unedited_data/Seurat_Breast_Bassez1_MINIATURE.rds"
  QCS: true
  qcs_save_file: "Seurat_Breast_Bassez1.rds"
  qcs_indent_var: "orig.ident"
  qcs_std_nFeature_RNA: 200
  qcs_vars_regress: ["nCount_RNA", "percent.mt"]
  qcs_ncells: 1
  qcs_nfeatures: 1
  ANNO: true
  anno_cutoff: 50
  anno_save_file: "Seurat_Breast_Bassez1_MINIATURE2.rds"
  DEx: true
  dex_cluster_algorithm: 1
  dex_save_file: "DE_bassez1_MINATURE2.csv"
  dex_stdev: 3
  dex_min_pct: 0.25
  dex_only_post: true ### If you want only positive, use this.
  dex_logfc_threshold: 0.0
  dex_p_val_adj: 0.1
---

```{r, echo=FALSE, warning=FALSE, message=FALSE, error=FALSE}
options(knitr.duplicate.label = 'allow')
knitr::opts_chunk$set(message = FALSE)
library(here)
```

# 1 - Quality control and standardization file (QCStanAnn)
Perform quality control and standardization of single cell RNA-seq datasets.

```{r child = 'QCS_01_05_2022.Rmd', eval=params$QCS}
```

# 2 - Annotation
Perform cell type annotation of single cell RNA-seq datasets.

```{r child = 'Anno_01_06_2022.Rmd', eval=params$ANNO}
```

# 3 - Differentilal Expression analysis
Identify differentially expressed genes in scRNA-seq data

```{r child = 'DEx_01_07_2022.Rmd', eval=params$DEx}
```

# Information about the Session 

Information about R, the OS and attached or loaded packages

```{r sesion_info, echo=FALSE}
pander::pander(sessionInfo(), compact = FALSE)
```
