---
title: "Single cell analysis: `r params$analysed_dataset`"
date: "Analysed on `r format(Sys.time(), '%B %d, %Y')`"
author: "QuantBio `r params$more_authors`"
output:
  rmdformats::downcute:
    self_contained: true
    thumbnails: false
    lightbox: true
    gallery: false
    highlight: tango
    code_folding: hide
    fig_width: 10
    fig_height: 10
    toc_depth: 5
params:
  analysed_dataset: "Chang_scRNASeq - Su_7142_210727B6 "
  more_authors: "Jennifer Modliszewski, Ph.D."
  raw_seurat: "../../Dropbox (QuantBio LLC)/QB_shared/Reports/Kirschlab/Chang_scRNA_Seq/processedData/Su_7142_210727B6_raw.rds"
  QCS: true
  qcs_prepare_raw_seurat: true
  qcs_mtx_dir: "./../cellranger_output"
  qcs_merge_sample_seurat: true
  qcs_save_file: "Seurat_QCS.rds"
  qcs_indent_var: "orig.ident"
  qcs_std_nFeature_RNA: 200
  qcs_vars_regress: ["nCount_RNA", "percent.mt"]
  qcs_ncells: 1
  qcs_nfeatures: 1
  qcs_patient_id: "patient_id"
  qcs_count_cutoff: "NA"
  qcs_feature_cutoff: "NA"
  qcs_mt_cutoff: "NA"
  ANNO: true
  anno_cutoff: 50
  anno_sigleR: true # Make sure qcs_save_file contains the ALSO the annotation with sigleR
  anno_Bcells: ["Pro-B_cell_CD34+","Pre-B_cell_CD34-","B_cell"]
  anno_HCS: ["HSC_-G-CSF","HSC_CD34+"]
  anno_save_file: "Seurat_Annotation.rds"
  anno_fibroblast: ["Fibroblasts", "Chondrocytes","Smooth_muscle_cells","Tissue_stem_cells"]
  anno_Others: ["Embryonic_stem_cells","GMP","Hepatocytes","iPS_cells","Keratinocytes","MSC","Neuroepithelial_cell","Neurons"]
  DEx: true
  dex_ident_cell_type: "SingleR2"
  dex_cluster_algorithm: 1
  dex_save_file: "Seurat_DEX.rds"
  dex_min_pct: 0.25
  dex_variance_explained: 0.8
  dex_only_post: true ### If you want only positive, use this.
  dex_logfc_threshold: 0.0
  dex_p_val_adj: 0.1
---

```{r, echo=FALSE, warning=FALSE, message=FALSE, error=FALSE}
options(knitr.duplicate.label = 'allow')
knitr::opts_chunk$set(message = FALSE, warning = FALSE, echo = TRUE,
                      tidy='styler', tidy.opts=list(strict=FALSE))
library(here)

theme_sara <- function(){
  theme_bw(base_size=14)+
    theme(axis.text=element_text(color="black"),
          panel.background=element_rect(color="white"),
          strip.text = element_text(size=12),
          strip.background = element_rect(fill="white"))
}

```

# 1 - Quality control and standardization file (QCStanAnn)
Perform quality control and standardization of single cell RNA-seq datasets.

```{r child = 'QCS_01_05_2022.Rmd', eval=params$QCS}
```

# 2 - Annotation
Perform cell type annotation of single cell RNA-seq datasets.

```{r child = 'Anno_01_06_2022.Rmd', eval=params$ANNO}
```

# 3 - Differentilal Expression analysis
Identify differentially expressed genes in scRNA-seq data

```{r child = 'DEx_01_07_2022.Rmd', eval=params$DEx}
```

# Information about the Session 

<details>
  <summary>Information about R, the OS and attached or loaded packages</summary>
```{r sesion_info}
pander::pander(sessionInfo(), compact = FALSE)
```
</details>
