---
editor_options: 
  markdown: 
    wrap: 72
---

```{r setup, include=FALSE, echo=FALSE}
### Set knitr global options
knitr::opts_chunk$set(warning=FALSE, message=FALSE)

### Fix conflicts
conflict_prefer("filter", "dplyr")
conflict_prefer("count", "dplyr")
conflict_prefer("select", "dplyr")
conflict_prefer("slice", "dplyr")
conflict_prefer("rename","dplyr")
conflict_prefer("mutate","dplyr")
conflict_prefer("arrange","dplyr")
conflict_prefer("summarize", "dplyr")
conflict_prefer("summarise", "dplyr")
conflict_prefer("layout","plotly")
```

> ***NOTE:*** Make sure the data is the unedited seurat file.

```{r load data}
###---------------------LOAD THE SEURAT OBJECT---------------------------------
if(params$is_list==TRUE & params$split_by_group==FALSE) {
    seurats <- readRDS(file=params$dir_seurat)
    
    for(i in 1:length(seurats)) {
        Idents(seurats[[i]]) <- seurats[[i]]@meta.data[,params$qcs_indent_var]
        seurats[[i]][["percent.mt"]] <- PercentageFeatureSet(object = seurats[[i]], pattern = "^MT|^mt")
    }
    
    } else {
        
    seurat_analysed <- readRDS(file=params$dir_seurat)
    
    # seurat_analysed@meta.data[,"orig.ident"] = gsub("BIOKEY", "Sample", seurat_analysed@meta.data[,"orig.ident"])
    # seurat_analysed@meta.data[,"Cell"] = gsub("BIOKEY", "Sample", seurat_analysed@meta.data[,"Cell"])
    # seurat_analysed@meta.data[,"patient_id"] = gsub("BIOKEY", "Sample", seurat_analysed@meta.data[,"patient_id"])

    ### Change ident to original idents
    Idents(seurat_analysed)<-seurat_analysed@meta.data[,params$qcs_indent_var]

    ###------------------ADD mito.gene % IN METADATA--------------------------------
    seurat_analysed[["percent.mt"]] <- PercentageFeatureSet(object = seurat_analysed, pattern = "^MT|^mt")
    
    ### Set identity
    if(params$group_name=="none"){
        seurat_analysed@meta.data$params_group<-"No group"
    } else {
        seurat_analysed@meta.data$params_group<-seurat_analysed@meta.data[,params$group_name]
    }
}


###---------FOR MULTIPLE BATCHES THAT NEED TO BE PROCESSED SEPARATELY----------
### This is often seen when CD45+ and CD45- are ran separately. The two extremely
### different cell types can have varying degrees of doublets and mitochondria
### gene expression. As such, they should be processed separately. If ran in a 
### single batch, then this is not necessary, but we'll still put it into a list.
if(params$is_list==FALSE & params$split_by_group==TRUE) {
    seurats <- SplitObject(seurat_analysed, split.by = params$group_name)
} 

if(params$is_list==FALSE & params$split_by_group==FALSE) {
    seurats<-list(seurat_analysed)
    names(seurats)<-params$group_name
}

###------------------DEFINE THE COLORS FOR REPORT------------------------------
colz<-list()
for(i in 1:length(seurats)){
    colz[[i]]<-get_colorz(length(unique(seurats[[i]]@meta.data[,params$qcs_indent_var])), 
                          palette1, 
                          unique(seurats[[i]]@meta.data[,params$qcs_indent_var]))
}
```

## 1.1 - Quality Control {.tabset .tabset-pils}

In section 1.1, we remove dead cells, doublets, and samples that are
flagged as outliers. In addition, we can remove genes that only show up
in a set number of cells. For example, if the gene only shows up in
1/100,000 cells, it is worth removing.

### Cell level (pre removal) {.tabset .tabset-pils}

Here we will remove dead cells, doublets, and cells with low molecule
counts. We do so by setting cut-offs for the following 3 parameters:

1.  `nFeature_RNA` is the number of genes detected in each cell.
2.  `nCount_RNA` is the total number of molecules detected within a cell
3.  `percent.mt` is the percent of features that are mitochondrial genes

**NOTE**

There are several things to consider. For one, a much higher number of
gene counts of molecule counts may indicate the presence of doublets,
especially for droplet-based scRNA-seq like 10X. In addition, high
mitochondrial gene percentages may indicate dead cells, which is common
in plate-based scRNA-seq like smart-seq.

```{r pre-qc-cells, results="asis", fig.align='center', fig.height=12, fig.width=10, eval=params$qcs_eval_cell}
### Plot the three parameters mentioned above
pre_plots<-list()
for (i in 1:length(seurats)) {
    p1 <- eval(substitute(
        
        plot_grid(
            
            violin_me(seurats[[i]], "nFeature_RNA",colorz = colz[[i]], themes$QB_theme_90_small(), legend = "none") + ggtitle("Gene count",subtitle=paste0(names(seurats[i]))),
            violin_me(seurats[[i]], "nCount_RNA",colorz = colz[[i]], themes$QB_theme_90_small(), legend="none")+ggtitle("UMI count",subtitle=paste0(names(seurats[i]))),
            violin_me(seurats[[i]], "percent.mt",colorz = colz[[i]], themes$QB_theme_90_small(), legend="none")+ggtitle("% MT genes",subtitle=paste0(names(seurats[i]))),
            
            ncol=1  ### For column count
            
            )

        ,list(i = i)))
    pre_plots[[i]] <- p1  # add each plot into plot list
}


### Plot it out
# cowplot::plot_grid(plotlist = pre_plots, nrow = length(seurats), ### Fix beginning here
#                    ncol = 1,
#                    vjust = 0.5)

#### Plot it
for (i in 1:length(pre_plots)){
print(glue::glue("#### {as.vector(names(seurats[i]))}\n"))
  
title_plots <- pre_plots[[i]]
print(title_plots)
cat("\n \n \n")
}


###-----------------------------CELL FILTER-------------------------------------
### Filter out cells that have QC metrics +/- away from 3*MAD:
### nFeature_RNA metric > 200-3*MAD(nFeature_RNA) and < median(nFeature_RNA)+3*MAD(nFeature_RNA)
### nCount_RNA < median(nCountRNA)+3*MAD(nCountRNA)
### percent.mt < median(percent.mt)+3*MAD(percent.mt)
for(i in 1:length(seurats)) {
    seurats[[i]] <- subset(x = seurats[[i]], 
                          subset = nFeature_RNA >  max(params$qcs_std_nFeature_RNA, median(seurats[[i]]@meta.data$nFeature_RNA, na.rm = TRUE) - 3*mad(seurats[[i]]@meta.data$nFeature_RNA, constant = 1.4826, na.rm = TRUE)) &
                              nFeature_RNA < median(seurats[[i]]@meta.data$nFeature_RNA, na.rm = TRUE) + 3*mad(seurats[[i]]@meta.data$nFeature_RNA, constant = 1.4826, na.rm = TRUE) &
                              nCount_RNA <  median(seurats[[i]]@meta.data$nCount_RNA, na.rm = TRUE) + 3*mad(seurats[[i]]@meta.data$nCount_RNA, constant = 1.4826, na.rm = TRUE) & 
                              percent.mt < median(seurats[[i]]@meta.data$percent.mt, na.rm = TRUE) + 3*mad(seurats[[i]]@meta.data$percent.mt, constant = 1.4826, na.rm = TRUE)
                  )
}
```

<br> <br>

### Cell level (post removal) {.tabset .tabset-pils}

Here we will remove dead cells, doublets, and cells with low molecule
counts. We do so by setting cut-offs for the following 3 parameters:

1.  `nFeature_RNA` is the number of genes detected in each cell.
2.  `nCount_RNA` is the total number of molecules detected within a cell
3.  `percent.mt` is the percent of features that are mitochondrial genes

**NOTE**

There are several things to consider. For one, a much higher number of
gene counts of molecule counts may indicate the presence of doublets,
especially for droplet-based scRNA-seq like 10X. In addition, high
mitochondrial gene percentages may indicate dead cells, which is common
in plate-based scRNA-seq like smart-seq.

```{r post-qc-cells, results="asis", fig.align='center', fig.height=12, fig.width=10, eval=params$qcs_eval_cell}
post_plots<-list()
for (i in 1:length(seurats)) {
    p1 <- eval(substitute(
        
        plot_grid(
            
            violin_me(seurats[[i]], "nFeature_RNA",colorz = colz[[i]], themes$QB_theme_90_small(), legend = "none") + ggtitle("Gene count",subtitle=paste0(names(seurats[i]))),
            violin_me(seurats[[i]], "nCount_RNA",colorz = colz[[i]], themes$QB_theme_90_small(), legend="none")+ggtitle("UMI count",subtitle=paste0(names(seurats[i]))),
            violin_me(seurats[[i]], "percent.mt",colorz = colz[[i]], themes$QB_theme_90_small(), legend="none")+ggtitle("% MT genes",subtitle=paste0(names(seurats[i]))),
            
            ncol=1  ### For column count
            
            )

        ,list(i = i)))
    post_plots[[i]] <- p1  # add each plot into plot list
}


#### Plot it
for (i in 1:length(post_plots)){
print(glue::glue("#### {as.vector(names(seurats[i]))}\n"))
  
title_plots <- post_plots[[i]]
print(title_plots)
cat("\n \n \n")
}
```

<br> <br>

### Gene level

Here we remove genes that only show up in a certain number of cells. You
can hit the "code" button to reveal the underlying code to remove low
count genes. There may be genes with zero counts across all cells too.
These reduce the average expression for a cell and should be removed, as
well. For our data we choose to keep only genes which are expressed in
**`r params$qcs_ncells` or more cells**.

<br>

This subsection is **code only**. Hit the "code" button to reveal the
code.

```{r remake-seurat, tidy='styler', tidy.opts=list(strict=FALSE)}
### Create the new seurat object
seurats <- lapply(X = seurats, FUN = function(x) {
    
    CreateSeuratObject(LayerData(x, assay="RNA", layer="counts"),
                       meta.data=x@meta.data,
                       min.cells=10,
                       min.features=10
        
    )
})
```

<br> <br>

### Sample level {.tabset .tabset-pils}

Here we remove samples that are flagged as outliers. This will remove
whole samples if the sample meets the following criteria:

-   Is sample **UMI median below 25% quantile** of the UMI count of the
    entire dataset?
-   Is sample **gene count below 25% quantile** of the gene count of the
    entire dataset?
-   Is sample ***mt.percent*** **above 75% quantile** of the mt.percent
    of whole dataset?

**NOTE**

We use quantiles as opposed to 3 x MAD because the bottom 3 x MAD will
go into a negative value, which does nothing.

```{r qc-samples, results="asis", eval=params$qcs_eval_sample, fig.height=12, fig.width=10}
### Plot the density of gene expression values by sample
m.melts<-list()
for(i in 1:length(seurats)){
    m.melts[[i]]<-seurats[[i]]@meta.data |>
        select(params$qcs_patient_id, nCount_RNA, nFeature_RNA, percent.mt) |>
        gather(key = "variable", value = "value", -params$qcs_patient_id) 
}

###---------------------STATISTIC CALCULATIONS----------------------------------
### Calculate some statistics to be used for plotting
count.meds<-lapply(X=m.melts, FUN=function(x) {
    
    ddply(x |> filter(variable=="nCount_RNA"),
          params$qcs_patient_id, summarise, med=median(value))    

})

feature.meds<-lapply(X=m.melts, FUN=function(x) {
    
    ddply(x |> filter(variable=="nFeature_RNA"),
          params$qcs_patient_id, summarise, med=median(value))    

})

mt.meds<-lapply(X=m.melts, FUN=function(x) {
    
    ddply(x |> filter(variable=="percent.mt"),
          params$qcs_patient_id, summarise, med=median(value))    

})

### Calculate overall medians
overall.stats<-list()
for(i in 1:length(m.melts)){
    overall.stats[[i]]<-m.melts[[i]] |> group_by(variable) |>
        summarize(median = median(value, na.rm = TRUE),
                  quantile = quantile(value, na.rm=TRUE))
    overall.stats[[i]]$quantile_name<-c(rep(c("0%","25%","50%","75%","100%"), 3))
    }

### Define variable for plotting, and the cutoffs to use
variables<-c("nCount_RNA","nFeature_RNA","percent.mt")
calculations<-c("count_quants","feature_quants","mt_quants")
cutoffs<-list()
for(i in 1:length(seurats)) {
    cutoffs[[i]]<-c(subset(overall.stats[[i]], variable=="nCount_RNA" & quantile_name == "25%")$quantile,
                    subset(overall.stats[[i]], variable=="nFeature_RNA" & quantile_name == "25%")$quantile,
                    subset(overall.stats[[i]], variable=="percent.mt" & quantile_name == "75%")$quantile)
    names(cutoffs[[i]])<-c("nCount_RNA","nFeature_RNA","percent.mt")
}
###-------------------------END STATISTICS CALC--------------------------------
### Plot the stuff
ggplotz<-list()
for(i in 1:length(m.melts)){
 
         p1 <- eval(substitute(
         plot_grid(

             
                sample_boxplot(dat= m.melts[[i]] |> filter(variable==variables[1]), varz=variables[1],  theme=themes$QB_theme_90_small(), colorz=colz[[i]],
                               line=cutoffs[[i]]["nCount_RNA"]),
                sample_boxplot(dat= m.melts[[i]] |> filter(variable==variables[2]), varz=variables[2],  theme=themes$QB_theme_90_small(), colorz=colz[[i]],
                               line=cutoffs[[i]]["nFeature_RNA"]),
                sample_boxplot(dat= m.melts[[i]] |> filter(variable==variables[3]), varz=variables[3],  theme=themes$QB_theme_90_small(), colorz=colz[[i]],
                               line=cutoffs[[i]]["percent.mt"]),
                ncol = 1
                
     ),
     list(i = i)))
         ggplotz[[i]] <- p1  # add each plot into plot list   
}

#### Plot it
for (i in 1:length(ggplotz)){
print(glue::glue("#### {as.vector(names(seurats[i]))}\n"))
  
title_plots <- ggplotz[[i]]
print(title_plots)
cat("\n \n \n")
}

```

<br> <br>

#### Removed samples (if any)

```{r, remove-sample-outliers}
### Remove samples that meet all 3 of the following criteria:
d<-list()
keepme<-list()
for(i in 1:length(count.meds)) {
    d[[i]]<-cbind(count.meds[[i]],feature.meds[[i]]$med, mt.meds[[i]]$med)
    colnames(d[[i]])<-c("Samples","Count_medians","Feature_medians", "Percent_mito")
    
    #### Check if any of the 4 variables are outliers
    d[[i]]$`Less than UMI count Q25`<- d[[i]]$Count_medians < ifelse(params$qcs_count_cutoff=="NA",
                                                        cutoffs[[i]]["nCount_RNA"],
                                                        params$qcs_count_cutoff)
    d[[i]]$`Less than gene count Q25`<-d[[i]]$Feature_medians< ifelse(params$qcs_count_cutoff=="NA",
                                                        cutoffs[[i]]["nFeature_RNA"],
                                                        params$qcs_feature_cutoff)
    d[[i]]$`Greater than %mt Q75`<-d[[i]]$Percent_mito > ifelse(params$qcs_count_cutoff=="NA",
                                                        cutoffs[[i]]["percent.mt"],
                                                        percent.mt_cutoff)

    ### Annotate the samples in which all 3 variables are outliers
    d[[i]]$remove_me<-d[[i]]$`Less than UMI count Q25`=="TRUE"&
        d[[i]]$`Less than gene count Q25`=="TRUE"&
        d[[i]]$`Greater than %mt Q75`=="TRUE"
    
    ### For annotation purposes
    d[[i]]$Group<-names(seurats[i])
    
    keepme[[i]]<-d[[i]]$Samples[which(d[[i]]$remove_me=="FALSE")]

}

### Unlist it for printing
stacked_dataframe <- do.call(rbind, d) |> 
    select(Group,
           Samples, 
           Count_medians, 
           Feature_medians, 
           Percent_mito,
           `Less than UMI count Q25`,
           `Less than gene count Q25`,
           `Greater than %mt Q75`,
           remove_me) |>
    rename(`Median UMI count`=Count_medians,
           `Median gene count`=Feature_medians,
           `Median %mt genes`=Percent_mito,
           `Selected for removal`=remove_me)

### Table it
stacked_dataframe |> render_dt()
```

<br> <br>

```{r, remove-sample-outliers}
### Remove from seurat
for (i in 1:length(seurats)){
    Idents(seurats[[i]])<-seurats[[i]]@meta.data[,params$qcs_patient_id]
    seurats[[i]]<-subset(seurats[[i]],idents=keepme[[i]])
}
```

<br> <br>

## 1.2 - Standardize {.tabset .tabset-pils}

In this section, we normalize the data, regress out confounding
variables, and scale the data. Each tab provides more detail.

### Normalize the data

This subsection is **code only**. Hit the "code" button to reveal the
code. For data normalization, we use `NormalizeDatas()` that normalizes
the feature expression measurements for each cell by the total
expression, multiplies this by a scale factor (10,000 by default), and
log-transforms the result. Normalized values are stored in
**[["RNA"]]@data**.

```{r normalize-date, tidy='styler', tidy.opts=list(strict=FALSE), eval=params$qcs_normalize}
seurats <- lapply(X=seurats, FUN=function(x) {
     NormalizeData(x,
                   normalization.method = "LogNormalize", 
                   scale.factor = 10000)   
})
```

<br> <br>

### Regress out confounding variables {.tabset .tabset-pils}

Calculate a subset of features that exhibit high cell-to-cell variation
in the dataset (i.e, they are highly expressed in some cells, and lowly
expressed in others). These genes help to highlight biological signal in
single-cell datasets.

By default, `FindVariableFeatures` returns 2,000 features per dataset,
but can be changed to any number. These may be used in downstream
analysis, like PCA.

**NOTE**

This will be done for each tissue type individually, so make sure to run
through them one at a time.

```{r plot-variable-features, results="asis", tidy='styler', tidy.opts=list(strict=FALSE), fig.align='center', fig.cap="Variable features with labels. Identiy the 10 most highly variable genes."}
### Find the top 2000 most variable genes for PCA analysis
seurats <- lapply(X=seurats, FUN=function(x) {
    FindVariableFeatures(object=x, selection.method = "vst", 
                            nfeatures = 2000)
})

### Plot variable features with labels
### Identify the 10 most highly variable genes
top10<-list()
labels<-list()
LabelPoints<-list()
for(i in 1:length(seurats)){
    top10[[i]] <- head(VariableFeatures(seurats[[i]]), 10)
    labels[[i]] <- VariableFeaturePlot(seurats[[i]])
    LabelPoints(plot = labels[[i]], points = top10[[i]], repel = TRUE)
}

### Generate a list of plots based on the above information
var_plots<-list()
for (i in 1:length(seurats)) {
    p1 <- eval(substitute(

                LabelPoints(plot = labels[[i]], points = top10[[i]], repel = TRUE)

        ,list(i = i)))
    var_plots[[i]] <- p1  # add each plot into plot list
}


#### Plot it out
for (i in 1:length(var_plots)){
print(glue::glue("#### {as.vector(names(seurats[i]))}\n"))
  
title_plots <- var_plots[[i]]
print(title_plots)
cat("\n \n \n")
}
```

### Scale the data

This subsection is **code only**. Hit the "code" button to reveal the
code. Next, we apply a linear transformation ('scaling') that is a
standard pre-processing step prior to dimensional reduction techniques
like PCA. The `ScaleData()` function:

1.  Shifts the expression of each gene, so that the mean expression
    across cells is 0
2.  Scales the expression of each gene, so that the variance across
    cells is 1. This step gives equal weight in downstream analyses, so
    that highly-expressed genes do not dominate
3.  The results of this are stored in **[["RNA"]]@scale.data**

**NOTE**

By default `ScaleData()` will only scale the identified 2,000 genes
unless specified using the `"features = all.genes"` argument.

```{r scale-data, tidy='styler', tidy.opts=list(strict=FALSE), eval=params$qcs_scale}
if (params$qcs_scale_genes=="all") {
  seurats <- lapply(X=seurats, FUN=function(x) {
      ScaleData(object = x,
                features = rownames(seurat_analysed), 
                verbose = FALSE,
                vars.to.regress = c(unlist(strsplit(params$qcs_vars_regress,
                                                  split = ",", fixed = T))))
        print("Scaled using all genes")
      }) 
} else {
  seurats <- lapply(X=seurats, FUN=function(x) {
      ScaleData(object = x,
                verbose = FALSE,
                vars.to.regress = c(unlist(strsplit(params$qcs_vars_regress,
                                                  split = ",", fixed = T))))
      }) 
  print("Scaled using top variable genes")
      
}
```

<br> <br>

## 1.3 - Identify the cell cycle phase for each cell

Here we infer the cell cycle phase in which the each cell is suspected
to be in based on gene expression levels of cell cycle markers.

### Score the cell cycle

This subsection is **code only**. Hit the "code" button to reveal the
code.

```{r cell-cycle-intro, include=FALSE, eval=params$qcs_cc}
qcs_cc_text = "This is done by calculating the G2M marker score with the `CellCycleScoring()` function provided in the _seurat package_."
if(!as.logical(params$qcs_cc)){
  qcs_cc_text = "> **_NOTE:_** This step has been skipped."
}
```

```{r cell-cycle-eval, eval=params$qcs_cc}
if (params$organism=="human") {
 ### [1] Create G2M marker gene list (i.e. genes associated with the G2M phase of the 
### cell cycle using Seurat's built-in cc.genes (cell cycle) genes list
g2m.genes <- cc.genes$g2m.genes
s.genes <- cc.genes$s.genes
g2m <- rownames(seurats[[1]])[rownames(seurats[[1]]) %in% g2m.genes]

} else if (params$organism=="mouse") {
  cc_file <- RCurl::getURL("https://raw.githubusercontent.com/hbc/tinyatlas/master/cell_cycle/Mus_musculus.csv") 
  cell_cycle_genes <- read.csv(text = cc_file)
  
  ah <- AnnotationHub::AnnotationHub()

  # Access the Ensembl database for organism
  ahDb <- AnnotationHub::query(ah, 
              pattern = c("Mus musculus", "EnsDb"), 
              ignore.case = TRUE)

  # Acquire the latest annotation files
  id <- ahDb %>%
          GenomicRanges::mcols() %>%
          rownames() %>%
          tail(n = 1)

  # Download the appropriate Ensembldb database
  edb <- ah[[id]]

  # Extract gene-level information from database
  annotations <- genes(edb, 
                       return.type = "data.frame")

  # Select annotations of interest
  annotations <- annotations %>%
                  dplyr::select(gene_id, gene_name, seq_name, gene_biotype, description)


  # Get gene names for Ensembl IDs for each gene
  cell_cycle_markers <- dplyr::left_join(cell_cycle_genes, annotations, by = c("geneID" = "gene_id"))

  # Acquire the S phase genes
  s.genes <- cell_cycle_markers %>%
              dplyr::filter(phase == "S") %>%
              dplyr::pull("gene_name")
        
  # Acquire the G2M phase genes        
  g2m.genes <- cell_cycle_markers %>%
                dplyr::filter(phase == "G2/M") %>%
                dplyr::pull("gene_name")
} 


### [2] Calculate G2M marker module score.
seurats <- lapply(X=seurats, FUN=function(x) {
    CellCycleScoring(x,
                     s.features = s.genes, ### Genes associated with s-phase
                     g2m.features = g2m.genes, ### Genes associated with G2M phase
                     set.ident = TRUE)
})
```

## 1.4 - Write to file

The QC is done, the file is save on ***`r paste0(gsub(".rds","",params$dir_seurat),params$qcs_save_name,".rds")`***.

```{r save-data, include=FALSE, eval=params$qcs_save}
### Merge it and write final object
write_rds(seurats, file=paste0(gsub(".rds","",params$dir_seurat),params$qcs_save_name,".rds"))
```
